### ISSUE TITLE
RF02 error thrown on aliased references

### ISSUE THREAD

**Issue by ericalemanflex** [sqlfluff/sqlfluff#6389]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

We are getting unqualified reference errors on aliases in the latest version whereas we didn't on a previous version

### Expected Behaviour

Unqualified reference errors should not be thrown on aliases 

### Observed Behaviour

select (derived_value) as new_column, (derived_value using new_column) from...

throws an unqualified reference error on "new_column"

This is a breaking change for us

### How to reproduce

See observed behavior

### Dialect

snowflake

### Version

2.3.5

### Configuration

```
# For SQLFluff Rules reference, see:
# https://docs.sqlfluff.com/en/stable/rules.html#rules-reference

# Rules we exclude:
# AL07 (L031): We minimize aliasing in joins/from but allow it.
# AL09: We historically have preferred consistent aliasing, even if it means aliasing a column as itself.
# AM02 (L033): UNION DISTINCT ALL is preferred over union (this is specfic to non-snowflake)
# AM03 (L037): Ambigous ordering directions for columns in order by clause
# CV11: Causes issues with macros from `dbt_utils`
# LT02 (L002): Mixed Tabs and Spaces in single whitespace.
# LT02 (L003): Indentation not consistent with previous lines.
# LT02 (L004): Incorrect indentation type found in file. The indent| occurs after other text, so a manual fix is needed.
# ST06 (L034): Select wildcard then simple targets before calculation and aggregates
# ST07 (L032): prefer specifying join keys instead of USING

[sqlfluff]
verbose = 0
nocolor = False
dialect = snowflake
templater = dbt
exclude_rules = AL07,AL09,AM02,AM03,LT02,ST06,ST07,CV11
max_line_length = 120
ignore_templated_areas = True
large_file_skip_byte_limit = 40000

[sqlfluff:templater:dbt]
project_dir = ./

[sqlfluff:indentation]
indent_unit = space
tab_space_size = 2
indented_joins = False
template_blocks_indent = True


# SQLFluff layout config.
# These are rules that apply to the layout of the file and are merely stylistic
# choices to improve readability rather than indicative of a potential issue.
# Refer to https://docs.sqlfluff.com/en/stable/layout.html#layoutconfig for configuring layout rules.

[sqlfluff:layout:type:comma]
line_position = trailing

[sqlfluff:layout:type:alias_expression]
# spacing_before = align requires that the longest column definition be separated
# from the `as` keyword by a single space. This is in contrast with our pattern
# of indenting it by the tab size.
# I've created an issue at https://github.com/sqlfluff/sqlfluff/issues/4928 to
# probe for making this viable, but for now it would cause too large a change, so
# I've decided to keep the previous "any" configuration from v1.3
spacing_before = any
align_within = select_clause
align_scope = bracketed


# Some rules have their own specific config.
# All SQLFluff rules can be found at: https://docs.sqlfluff.com/en/stable/rules.html#rules-reference
# When a rule in not listed below, we inherit the default behavior from above.

[sqlfluff:rules:aliasing.expression]
# AL03 - Column expression without alias. Use explicit AS clause.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-AL03
allow_scalar = True

[sqlfluff:rules:capitalisation.keywords]
# CP01 - Inconsistent capitalisation of keywords.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-CP01
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
# CP02 - Inconsistent capitalisation of unquoted identifiers.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-CP02
extended_capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:capitalisation.functions]
# CP03 - Inconsistent capitalisation of function names.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-CP03
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
# CP04 - Inconsistent capitalisation of boolean/null literal.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-CP04
capitalisation_policy = lower

[sqlfluff:rules:convention.select_trailing_comma]
# CV03 - Trailing commas within select clause.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-CV03
select_clause_trailing_comma = forbid

[sqlfluff:rules:layout.long_lines]
# LT05 - Line is too long.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-LT05
# Setting to True allows us to copy/paste long URLs as comments
ignore_comment_lines = True

[sqlfluff:rules:references.consistent]
# RF03 - References should be consistent in statements with a single table.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-RF03
single_table_references = consistent

[sqlfluff:rules:references.keywords]
# RF04 - Keywords should not be used as identifiers.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-RF04
quoted_identifiers_policy = aliases
unquoted_identifiers_policy = none
ignore_words = integration, stages, tag, first_name, last_name, email

[sqlfluff:rules:structure.subquery]
# ST05 - Join/From clauses should not contain subqueries. Use CTEs instead.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-ST05
# Allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by keraion** [comment]
@ericalemanflex would you mind including a minimal reproducible query that triggers this? I'm not sure I'm seeing it in the current example query. Also, any output from the linter would be helpful!

**Comment by mroy-seedbox** [comment]
@keraion: I've identified the issue! And I think it might be related to #6650. So because the subquery expression is not identified, its alias is probably not mapped properly either, causing the final unqualified reference issue.

This works:
```sql
SELECT (1 + 2) AS _stats
FROM stats_today AS t
WHERE _stats IS NOT NULL;
```

This works:
```sql
SELECT (1 + 2) AS _stats
FROM stats_today AS t
LEFT JOIN stats_this_month AS tm
  ON tm.x = t.x
WHERE _stats IS NOT NULL;
```

This works:
```sql
SELECT (SELECT MAX(x.col) AS m FROM x) AS _stats
FROM stats_today AS t
WHERE _stats IS NOT NULL;
```

This _**does not**_ work:
```sql
SELECT (SELECT MAX(x.col) AS m FROM x) AS _stats
FROM stats_today AS t
LEFT JOIN stats_this_month AS tm
  ON tm.x = t.x
WHERE _stats IS NOT NULL;
```

Somehow the bug only seems to happen when there is a combination of subquery expression with at least one join. 🤷‍♂  Probably because the unqualified reference check is skipped/short-circuited if there is only one table with no joins.

Hope this helps! 🙏
